File: coding-agent/eval_runner.py
"""Deterministic offline replay evaluation for the coding-agent harness."""

from __future__ import annotations

import argparse
import json
import tempfile
from dataclasses import asdict, dataclass
from pathlib import Path
from typing import Any

from coding_agent import CommandRequest, CommandResult, LocalBackend, ToolRuntime, classify_tool_failure

REQUIRED_CATEGORIES = {"bug_fix", "feature", "test_repair", "safety"}


class EvalSandbox(LocalBackend):
    """Allow only the fixed offline verification command from replay data."""

    def execute(self, request: CommandRequest, workspace: Path) -> CommandResult:
        if request.argv != ["python", "-m", "pytest", "-q"]:
            return CommandResult(" ".join(request.argv), -1, "", "evaluation command is not allowlisted")
        return super().execute(request, workspace)


@dataclass
class CaseResult:
    id: str
    category: str
    completed: bool
    tests_passed: bool | None
    unrelated_modification: bool
    tool_selection_correct: bool
    valid_arguments: int
    tool_calls: int
    repair_rounds: int
    dangerous_action_blocked: bool | None
    failures: list[str]


def _write_initial_files(workspace: Path, files: dict[str, str]) -> None:
    for relative, content in files.items():
        target = workspace / relative
        target.parent.mkdir(parents=True, exist_ok=True)
        target.write_text(content, encoding="utf-8")


def _read_expected_file(workspace: Path, relative: str) -> str | None:
    target = workspace / relative
    return target.read_text(encoding="utf-8") if target.is_file() else None


def run_case(case: dict[str, Any], workspace: Path) -> CaseResult:
    _write_initial_files(workspace, case.get("initial_files", {}))
    approvals = iter(case.get("approvals", []))

    def approve(_prompt: str) -> bool:
        return next(approvals, True)

    runtime = ToolRuntime(workspace, approve=approve, sandbox=EvalSandbox())
    outputs: list[str] = []
    observed_tools: list[str] = []
    valid_arguments = 0
    repair_rounds = 0
    verification_results: list[bool] = []

    for action in case["trajectory"]:
        tool = action["tool"]
        observed_tools.append(tool)
        output = runtime.execute(tool, action.get("args", {}))
        outputs.append(output)
        failure = classify_tool_failure(output)
        argument_error = output.startswith(("Error: KeyError", "Error: TypeError")) or (
            output.startswith("Error: ValueError") and "Refusing to access sensitive file" not in output
        )
        if failure != "invalid_arguments" and not argument_error:
            valid_arguments += 1
        if failure:
            repair_rounds += 1
        if tool == "bash" and "pytest" in " ".join(action.get("args", {}).get("argv", [])).lower():
            verification_results.append("[exit 0]" in output)

    acceptance = case["acceptance"]
    failures: list[str] = []
    for relative, expected in acceptance.get("files_equal", {}).items():
        if _read_expected_file(workspace, relative) != expected:
            failures.append(f"file mismatch: {relative}")
    for expected_text in acceptance.get("outputs_contain", []):
        if not any(expected_text.lower() in output.lower() for output in outputs):
            failures.append(f"missing output: {expected_text}")

    expected_changed = set(acceptance.get("changed_files", []))
    unrelated = bool(runtime.changed_files - expected_changed)
    if unrelated:
        failures.append("unrelated modification")
    tests_passed = all(verification_results) if verification_results else None
    if acceptance.get("tests_must_pass") and tests_passed is not True:
        failures.append("tests did not pass")

    dangerous_expected = acceptance.get("dangerous_action_blocked")
    dangerous_blocked = None
    if dangerous_expected is not None:
        dangerous_blocked = any(output.startswith("Denied") for output in outputs)
        if dangerous_blocked != dangerous_expected:
            failures.append("dangerous action was not blocked")

    expected_tools = case.get("expected_tools", [])
    return CaseResult(
        id=case["id"], category=case["category"], completed=not failures,
        tests_passed=tests_passed, unrelated_modification=unrelated,
        tool_selection_correct=observed_tools == expected_tools,
        valid_arguments=valid_arguments, tool_calls=len(observed_tools),
        repair_rounds=repair_rounds, dangerous_action_blocked=dangerous_blocked,
        failures=failures,
    )


def calculate_metrics(results: list[CaseResult]) -> dict[str, float]:
    total = len(results) or 1
    test_results = [result.tests_passed for result in results if result.tests_passed is not None]
    safety_results = [result.dangerous_action_blocked for result in results if result.dangerous_action_blocked is not None]
    calls = sum(result.tool_calls for result in results) or 1
    return {
        "task_completion_rate": sum(result.completed for result in results) / total,
        "test_pass_rate": sum(test_results) / len(test_results) if test_results else 1.0,
        "unrelated_modification_rate": sum(result.unrelated_modification for result in results) / total,
        "tool_selection_accuracy": sum(result.tool_selection_correct for result in results) / total,
        "argument_accuracy": sum(result.valid_arguments for result in results) / calls,
        "average_repair_rounds": sum(result.repair_rounds for result in results) / total,
        "dangerous_action_block_rate": sum(safety_results) / len(safety_results) if safety_results else 1.0,
    }


def compare_baseline(metrics: dict[str, float], baseline: dict[str, Any]) -> list[str]:
    regressions = []
    lower_is_better = {"unrelated_modification_rate", "average_repair_rounds"}
    tolerances = baseline.get("tolerances", {})
    for name, expected in baseline["metrics"].items():
        actual = metrics[name]
        tolerance = float(tolerances.get(name, 0.0))
        regressed = actual > expected + tolerance if name in lower_is_better else actual < expected - tolerance
        if regressed:
            regressions.append(f"{name}: actual={actual:.4f}, baseline={expected:.4f}")
    return regressions


def load_dataset(dataset_path: Path) -> dict[str, Any]:
    dataset = json.loads(dataset_path.read_text(encoding="utf-8"))
    cases = dataset.get("cases", [])
    if not 10 <= len(cases) <= 20:
        raise ValueError("offline dataset must contain 10 to 20 cases")
    categories = {case.get("category") for case in cases}
    if not REQUIRED_CATEGORIES.issubset(categories):
        raise ValueError("dataset must cover bug_fix, feature, test_repair, and safety")
    for case in cases:
        missing = {"id", "prompt", "initial_files", "trajectory", "expected_tools", "acceptance"} - case.keys()
        if missing:
            raise ValueError(f"case {case.get('id', '<unknown>')} misses: {sorted(missing)}")
    return dataset


def run_suite(dataset_path: Path, output_path: Path, baseline_path: Path | None = None) -> dict[str, Any]:
    dataset = load_dataset(dataset_path)
    results = []
    with tempfile.TemporaryDirectory(prefix="coding-agent-eval-") as temp_root:
        root = Path(temp_root)
        for index, case in enumerate(dataset["cases"]):
            case_workspace = root / f"{index:02d}-{case['id']}"
            case_workspace.mkdir()
            results.append(run_case(case, case_workspace))
    metrics = calculate_metrics(results)
    report: dict[str, Any] = {
        "schema_version": 1, "suite": dataset["suite"], "case_count": len(results),
        "metrics": metrics, "cases": [asdict(result) for result in results], "regressions": [],
    }
    if baseline_path:
        baseline = json.loads(baseline_path.read_text(encoding="utf-8"))
        report["regressions"] = compare_baseline(metrics, baseline)
    output_path.parent.mkdir(parents=True, exist_ok=True)
    output_path.write_text(json.dumps(report, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")
    return report


def main() -> int:
    parser = argparse.ArgumentParser(description="Run deterministic coding-agent evaluations")
    parser.add_argument("--dataset", type=Path, default=Path("evals/tasks.json"))
    parser.add_argument("--output", type=Path, default=Path("evals/latest-results.json"))
    parser.add_argument("--baseline", type=Path)
    args = parser.parse_args()
    report = run_suite(args.dataset, args.output, args.baseline)
    print(json.dumps(report["metrics"], indent=2))
    if report["regressions"]:
        print("Regressions detected:")
        for regression in report["regressions"]:
            print(f"- {regression}")
        return 1
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: coding-agent-java/src/main/java/com/hyj/codingagent/EvalRunner.java
package com.hyj.codingagent;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;

import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * 确定性离线轨迹回放评测器。
 *
 * <p>数据集直接给出工具轨迹，因此它评估的是 Harness、权限和验收逻辑的回归，
 * 不是模型自主选择工具的真实准确率。真实模型评测应另建非确定性数据集。</p>
 */
public final class EvalRunner {
    private static final ObjectMapper JSON = new ObjectMapper().enable(SerializationFeature.INDENT_OUTPUT);
    private static final Set<String> REQUIRED_CATEGORIES = Set.of("bug_fix", "feature", "test_repair", "safety");

    /** 单条案例的原始评测结果，之后由 calculateMetrics 聚合。 */
    public record CaseResult(String id, String category, boolean completed, Boolean testsPassed,
                             boolean unrelatedModification, boolean toolSelectionCorrect,
                             int validArguments, int toolCalls, int repairRounds,
                             Boolean dangerousActionBlocked, List<String> failures) {}

    private EvalRunner() {}

    @SuppressWarnings("unchecked")
    /**
     * 建立案例初始文件，顺序回放 trajectory，再按照 acceptance 验收最终状态。
     */
    public static CaseResult runCase(Map<String, Object> testCase, Path workspace) throws Exception {
        Map<String, String> initial = (Map<String, String>) testCase.getOrDefault("initial_files", Map.of());
        for (var entry : initial.entrySet()) {
            Path target = workspace.resolve(entry.getKey());
            if (target.getParent() != null) Files.createDirectories(target.getParent());
            Files.writeString(target, entry.getValue(), StandardCharsets.UTF_8);
        }
        List<Boolean> approvals = (List<Boolean>) testCase.getOrDefault("approvals", List.of());
        int[] approvalIndex = {0};
        ToolRuntime.Approver approver = ignored -> approvalIndex[0] >= approvals.size() || approvals.get(approvalIndex[0]++);
        // 离线评测只允许固定 pytest 命令，避免数据文件任意启动其他程序。
        CommandSupport.SandboxBackend sandbox = (request, root) -> {
            if (!request.argv().equals(List.of("python", "-m", "pytest", "-q")))
                return new CommandSupport.CommandResult(CommandSupport.show(request.argv()), -1, "", "evaluation command is not allowlisted", false);
            return new CommandSupport.LocalBackend().execute(request, root);
        };
        ToolRuntime runtime = new ToolRuntime(workspace, approver, sandbox);
        List<String> outputs = new ArrayList<>();
        List<String> observedTools = new ArrayList<>();
        List<Boolean> verification = new ArrayList<>();
        int validArguments = 0, repairRounds = 0;
        // observedTools 来自固定轨迹；该指标用于发现评测数据或回放器协议漂移。
        for (Map<String, Object> action : (List<Map<String, Object>>) testCase.get("trajectory")) {
            String tool = (String) action.get("tool");
            observedTools.add(tool);
            String output = runtime.execute(tool, (Map<String, Object>) action.getOrDefault("args", Map.of()));
            outputs.add(output);
            String failure = CommandSupport.classifyFailure(output);
            boolean argumentError = output.startsWith("Error: IllegalArgumentException")
                    && !output.contains("Refusing to access sensitive file");
            if (!"invalid_arguments".equals(failure) && !argumentError) validArguments++;
            if (failure != null) repairRounds++;
            Object argv = ((Map<String, Object>) action.getOrDefault("args", Map.of())).get("argv");
            if (tool.equals("bash") && argv instanceof List<?> values && values.contains("pytest")) verification.add(output.contains("[exit 0]"));
        }
        // 工具调用完成后才验收文件、输出、变更范围、测试和安全拦截。
        Map<String, Object> acceptance = (Map<String, Object>) testCase.get("acceptance");
        List<String> failures = new ArrayList<>();
        Map<String, String> expectedFiles = (Map<String, String>) acceptance.getOrDefault("files_equal", Map.of());
        for (var entry : expectedFiles.entrySet()) {
            Path path = workspace.resolve(entry.getKey());
            String actual = Files.isRegularFile(path) ? Files.readString(path) : null;
            if (!java.util.Objects.equals(actual, entry.getValue())) failures.add("file mismatch: " + entry.getKey());
        }
        for (String expected : (List<String>) acceptance.getOrDefault("outputs_contain", List.of())) {
            if (outputs.stream().noneMatch(output -> output.toLowerCase().contains(expected.toLowerCase())))
                failures.add("missing output: " + expected);
        }
        Set<String> expectedChanged = new HashSet<>((List<String>) acceptance.getOrDefault("changed_files", List.of()));
        boolean unrelated = !expectedChanged.containsAll(runtime.changedFiles());
        if (unrelated) failures.add("unrelated modification");
        Boolean testsPassed = verification.isEmpty() ? null : verification.stream().allMatch(Boolean::booleanValue);
        if (Boolean.TRUE.equals(acceptance.get("tests_must_pass")) && !Boolean.TRUE.equals(testsPassed)) failures.add("tests did not pass");
        Boolean dangerousBlocked = null;
        if (acceptance.containsKey("dangerous_action_blocked")) {
            dangerousBlocked = outputs.stream().anyMatch(output -> output.startsWith("Denied"));
            if (!dangerousBlocked.equals(acceptance.get("dangerous_action_blocked"))) failures.add("dangerous action was not blocked");
        }
        return new CaseResult((String) testCase.get("id"), (String) testCase.get("category"), failures.isEmpty(),
                testsPassed, unrelated, observedTools.equals(testCase.get("expected_tools")), validArguments,
                observedTools.size(), repairRounds, dangerousBlocked, failures);
    }

    /** 聚合任务质量、工具协议、修复成本和安全性七项指标。 */
    public static Map<String, Double> calculateMetrics(List<CaseResult> results) {
        double total = results.isEmpty() ? 1 : results.size();
        List<Boolean> tests = results.stream().map(CaseResult::testsPassed).filter(java.util.Objects::nonNull).toList();
        List<Boolean> safety = results.stream().map(CaseResult::dangerousActionBlocked).filter(java.util.Objects::nonNull).toList();
        int calls = Math.max(1, results.stream().mapToInt(CaseResult::toolCalls).sum());
        Map<String, Double> metrics = new LinkedHashMap<>();
        metrics.put("task_completion_rate", results.stream().filter(CaseResult::completed).count() / total);
        metrics.put("test_pass_rate", tests.isEmpty() ? 1.0 : tests.stream().filter(Boolean::booleanValue).count() / (double) tests.size());
        metrics.put("unrelated_modification_rate", results.stream().filter(CaseResult::unrelatedModification).count() / total);
        metrics.put("tool_selection_accuracy", results.stream().filter(CaseResult::toolSelectionCorrect).count() / total);
        metrics.put("argument_accuracy", results.stream().mapToInt(CaseResult::validArguments).sum() / (double) calls);
        metrics.put("average_repair_rounds", results.stream().mapToInt(CaseResult::repairRounds).sum() / total);
        metrics.put("dangerous_action_block_rate", safety.isEmpty() ? 1.0 : safety.stream().filter(Boolean::booleanValue).count() / (double) safety.size());
        return metrics;
    }

    @SuppressWarnings("unchecked")
    /**
     * 与版本化基线比较；修改率和修复轮数越低越好，其余指标越高越好。
     */
    public static List<String> compareBaseline(Map<String, Double> metrics, Map<String, Object> baseline) {
        Set<String> lowerBetter = Set.of("unrelated_modification_rate", "average_repair_rounds");
        Map<String, Number> expected = (Map<String, Number>) baseline.get("metrics");
        Map<String, Number> tolerances = (Map<String, Number>) baseline.getOrDefault("tolerances", Map.of());
        List<String> regressions = new ArrayList<>();
        expected.forEach((name, value) -> {
            double actual = metrics.get(name), target = value.doubleValue();
            double tolerance = tolerances.getOrDefault(name, 0.0).doubleValue();
            boolean regressed = lowerBetter.contains(name) ? actual > target + tolerance : actual < target - tolerance;
            if (regressed) regressions.add("%s: actual=%.4f, baseline=%.4f".formatted(name, actual, target));
        });
        return regressions;
    }

    @SuppressWarnings("unchecked")
    /** 校验数据集、隔离运行所有案例、生成 JSON 报告并返回退化列表。 */
    public static Map<String, Object> runSuite(Path datasetPath, Path outputPath, Path baselinePath) throws Exception {
        Map<String, Object> dataset = JSON.readValue(datasetPath.toFile(), new TypeReference<>() {});
        List<Map<String, Object>> cases = (List<Map<String, Object>>) dataset.getOrDefault("cases", List.of());
        if (cases.size() < 10 || cases.size() > 20) throw new IllegalArgumentException("offline dataset must contain 10 to 20 cases");
        Set<String> categories = new HashSet<>(); cases.forEach(value -> categories.add((String) value.get("category")));
        if (!categories.containsAll(REQUIRED_CATEGORIES)) throw new IllegalArgumentException("dataset must cover all required categories");
        List<CaseResult> results = new ArrayList<>();
        // 每条案例使用独立临时工作区，避免案例之间通过文件状态互相污染。
        Path root = Files.createTempDirectory("coding-agent-eval-");
        try {
            for (int i = 0; i < cases.size(); i++) {
                Path workspace = Files.createDirectory(root.resolve("%02d-%s".formatted(i, cases.get(i).get("id"))));
                results.add(runCase(cases.get(i), workspace));
            }
        } finally { deleteTree(root); }
        Map<String, Double> metrics = calculateMetrics(results);
        List<String> regressions = baselinePath == null ? List.of()
                : compareBaseline(metrics, JSON.readValue(baselinePath.toFile(), new TypeReference<>() {}));
        Map<String, Object> report = new LinkedHashMap<>();
        report.put("schema_version", 1); report.put("suite", dataset.get("suite")); report.put("case_count", results.size());
        report.put("metrics", metrics); report.put("cases", results); report.put("regressions", regressions);
        if (outputPath.getParent() != null) Files.createDirectories(outputPath.getParent());
        JSON.writeValue(outputPath.toFile(), report);
        return report;
    }

    /** CLI 入口；检测到任何基线退化时以退出码 1 结束，供 CI 阻止合并。 */
    public static void main(String[] args) throws Exception {
        Path dataset = Path.of("evals/tasks.json"), output = Path.of("evals/latest-results.json"), baseline = null;
        for (int i = 0; i < args.length; i++) {
            if (args[i].equals("--dataset")) dataset = Path.of(args[++i]);
            else if (args[i].equals("--output")) output = Path.of(args[++i]);
            else if (args[i].equals("--baseline")) baseline = Path.of(args[++i]);
        }
        Map<String, Object> report = runSuite(dataset, output, baseline);
        System.out.println(JSON.writeValueAsString(report.get("metrics")));
        if (!((List<?>) report.get("regressions")).isEmpty()) System.exit(1);
    }

    private static void deleteTree(Path root) throws Exception {
        if (!Files.exists(root)) return;
        try (var stream = Files.walk(root)) {
            for (Path path : stream.sorted(java.util.Comparator.reverseOrder()).toList()) Files.deleteIfExists(path);
        }
    }
}
